asllib/aslspec/spec.ml

```diff
@@ -448,16 +448,284 @@ module Check = struct
-        records match their definition (see [tests/type_instance.bad]). *)
+  (** * A module for conservatively checking that all type terms are
+      well-formed. That is each type term that instantiates another defined type
+      term is subsumed by it. *)
+  module CheckTypeInstantiations : sig
+    val check : definition_node StringMap.t -> elem list -> unit
+    (** [check id_to_defining_node elems] conservatively checks that all type
+        terms referenced in [elems] that instantiate type terms in the range of
+        [id_to_defining_node] are also subsumed by them. The check assumes that
+        [check_no_undefined_ids] has already been run. *)
+  end = struct
+    (** [subsumed sub_op super_op] is true if all values in the domain of
+        [Operator { op = sub_op; term }] are also in the domain of
+        [Operator { op = super_op; term }]. *)
+    let operator_subsumed sub_op super_op =
+      match (sub_op, super_op) with
+      | Powerset, Powerset
+      | Powerset_Finite, (Powerset_Finite | Powerset)
+      | List1, (List1 | List0)
+      | List0, List0
+      | Option, (Option | Powerset_Finite | Powerset)
+      (* Option represents sets with 0 or 1 values while Powerset represents sets
+         of _any_ number of values. *)
+        ->
+          true
+      | _ -> false
+
+    (** [subsumed id_to_defining_node expanded_types sub super] conservatively
+        tests whether all values in the domain of [sub] are also in the domain
+        of [super]. Labels that represent types may be expanded, that is,
+        replaced by their list of type variants, using [id_to_defining_node]. To
+        ensure termination on recursive types, this expansion is done at most
+        once by tracking the set of expanded labels in [expanded_types]. *)
+    let rec subsumed id_to_defining_node expanded_types sub super =
+      (* The domain of an unlabelled singleton tuple is the domain of its single type term. *)
+      let equiv_singleton_tuple term =
+        match term with
+        | LabelledTuple
+            { label_opt = None; components = [ (_, referenced_term) ] } ->
+            referenced_term
+        | _ -> term
+      in
+      let sub = equiv_singleton_tuple sub in
+      let super = equiv_singleton_tuple super in
+      match (sub, super) with
+      | _, Label super_label ->
+          let sub_is_label_case =
+            match sub with
+            | Label sub_label ->
+                String.equal sub_label super_label
+                ||
+                (* The case where [sub_label] is a type name. *)
+                subsumed_typename_term_type id_to_defining_node expanded_types
+                  sub_label super
+            | _ -> false
+          in
+          sub_is_label_case
+          (* The case where [super_label] is a type name. *)
+          || subsumed_term_type_typename id_to_defining_node expanded_types sub
+               super_label
```

Comment:
Could we document a case where recursing on super (resp sub) before sub (resp super) makes type-checking fail?